### PR TITLE
Improve memory utilization for Applications

### LINKED ISSUE #2401: Memory Spikes happen when an event received from message broker.
### Describe your problem(s)

When there are massive number of applications/application key mappings or APIs, memory spikes happen upon receiving an event from message broker. 

### Describe your solution

This is because the adapter maintains two types of instances for same instance representation; one for xds services and the other for adapter internal processing. This step is required as then the instance type declared in proto file will not be tightly coupled with the one the adapter uses for parsing, validation etc. 

The maps are maintained to handle instance addition/deletion and there are currently based on that adapter internal representation. Hence, it is required to construct xds representing instances for all items within these maps even for a single instance update. (ex. Application Update event is received and the whole set of types.application instances needs to be converted to subscription.Application)

To optimize the memory utlilization, the maps needs to be based on xds representing instances as it cuts off the requirement of constructing large number of instances repeatedly.

### How will you implement it

---

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose

Prevent memory usage spikes when an event is received from message broker. 

See the issue for more details.

### Issues

Fixes #2401 

### Automation tests
 - Unit tests added: No
 - Integration tests added: No (Already covered with available set of tests)

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: adapter/internal/eventhub/subscription.go [hunk 1]
```diff
@@ -166,7 +153,7 @@ func LoadSubscriptionData(configFile *config.Config, initialAPIUUIDListMap map[s
 	for _, configuredEnv := range configuredEnvs {
 		queryParamMap := make(map[string]string, 1)
 		queryParamMap[GatewayLabelParam] = configuredEnv
-		go InvokeService(ApisEndpoint, APIListMap[configuredEnv], queryParamMap, APIListChannel, 0)
+		go InvokeService(ApisEndpoint, apiList, queryParamMap, APIListChannel, 0)
```
Comment (VirajSalaka) [inline adapter/internal/eventhub/subscription.go:156]: double check if having a common apiList reference for all the env based calls.

Comment (VirajSalaka) [reply adapter/internal/eventhub/subscription.go:158]: InvokeService just uses the type of the apiList variable via reflect (to parse the response received from control plane).

File: adapter/internal/messaging/notification_listener.go [hunk 2]
```diff
@@ -173,37 +177,17 @@ func handleAPIEvents(data []byte, eventType string) {
...
-				apiListOfEnv := eh.APIListMap[env].List
-				for i := range apiListOfEnv {
-					// TODO: (VirajSalaka) Use APIId once it is fixed from control plane
-					if apiEvent.Context == apiListOfEnv[i].Context && apiEvent.Version == apiListOfEnv[i].Version {
-						eh.APIListMap[env].List = DeleteAPIFromList(apiListOfEnv, i, apiEvent.UUID, env)
-						xds.UpdateEnforcerAPIList(env, xds.MarshalAPIList(eh.APIListMap[env]))
-						break
+					if xds.CheckIfAPIMetadataIsAlreadyAvailable(apiEvent.UUID, env) {
+						logger.LoggerInternalMsg.Debugf("API Metadata for api Id: %s is not updated as it already exists", apiEvent.UUID)
+						continue
```
[the remark this answers is not in the document]
Comment (VirajSalaka) [reply adapter/internal/messaging/notification_listener.go:184]: This concern is tracked in a separate issue https://github.com/wso2/product-microgateway/issues/2425

**Review by pubudu538 — APPROVED** [review]

File: adapter/internal/discovery/xds/marshaller.go [hunk 3]
```diff
@@ -380,3 +328,326 @@ func MarshalKeyManager(keyManager *types.KeyManager) *keymgt.KeyManagerConfig {
...
+		ApplicationKeyMappingMap[applicationKeyMappingReference] = keyMappingSub
+		logger.LoggerXds.Infof("Application Key Mapping for the applicationKeyMappingReference %s is added.",
+			applicationKeyMappingReference)
+	}
+	updateEnforcerApplicationKeyMappings(marshalKeyMappingMapToList(ApplicationKeyMappingMap))
+}
+
+// UpdateXdsForMultipleSubscriptions is used to update the subscriptions during the startup where
+// multiple subscriptions are pulled at once. In the end, XDS cache will be updated as well.
+func UpdateXdsForMultipleSubscriptions(subscriptionsList *types.SubscriptionList) {
```
Comment (praminda) [inline adapter/internal/discovery/xds/marshaller.go:394]: better to use this file only for "marshalling" related functions. basically here you should only marshall the required objects and return. another party should call the update xds methods.  updating xds shouldn't be a part of marshalling logic's responsibility. Same for all other new `UpdateXdsFor...` methods introduced in here

File: adapter/internal/ga/api_event_listener.go [hunk 4]
```diff
@@ -35,44 +35,27 @@ func handleAPIEventsFromGA(channel chan APIEvent) {
...
+			if xds.CheckIfAPIMetadataIsAlreadyAvailable(event.APIUUID, env) {
+				logger.LoggerGA.Debugf("APIList for API UUID: %s is not updated as it already "+
+					"exists", event.APIUUID)
+				continue
 			}
+			queryParamMap := make(map[string]string, 2)
+			queryParamMap[eh.GatewayLabelParam] = env
+			queryParamMap[eh.APIUUIDParam] = event.APIUUID
+			logger.LoggerGA.Infof("Invoking the apis service endpoint")
+			// TODO: (VirajSalaka) fix
```
[the remark this answers is not in the document]
Comment (VirajSalaka) [reply adapter/internal/ga/api_event_listener.go]: yeah
https://github.com/wso2/product-microgateway/pull/2391#discussion_r748999199

**Review by praminda — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] dev-deployment : 20211202.25

**Pull request merged**